pkg/sql/catalog/tabledesc/table.go

```diff
@@ -145,13 +146,22 @@ func MakeColumnDefDescs(
 	col.Type = resType
 
 	if d.HasDefaultExpr() {
+		var innerErr error
 		// Verify the default expression type is compatible with the column type
 		// and does not contain invalid functions.
 		ret.DefaultExpr, err = schemaexpr.SanitizeVarFreeExpr(
 			ctx, d.DefaultExpr.Expr, resType, "DEFAULT", semaCtx, volatility.Volatile,
 		)
 		if err != nil {
-			return nil, err
+			// Check if the default expression type can be assignment-cast into the
+			// column type. If it can allow the default and column type to differ.
+			ret.DefaultExpr, innerErr = d.DefaultExpr.Expr.TypeCheck(ctx, semaCtx, types.Any)
```

Comment:
`schemaexpr.SanitizeVarFreeExpr` above could return an error for reasons other than the expression not matching the column type. For example, it could fail if the expression contains references to other columns. We only want to check if the default expression type can be assignment-cast into the column type if `schemaexpr.SanitizeVarFreeExpr` fails during type-checking. Unfortunately, this isn't easy to do from outside `schemaexpr.SanitizeVarFreeExpr`. One option is to modify `schemaexpr.SanitizeVarFreeExpr` to add a new argument, `allowAssignmentCast bool` and move his assignment-cast logic into the function directly where type-checking fails.

I think you could add a test case for this too, like:

```
CREATE TABLE fail_assn_cast (
  a BOOL,
  b STRING DEFAULT a
)
```